### PR TITLE
V1.3 feature flaw `EditCommandParser` should return stacked edit error messages, refactor `AddCommandParser`, add more tests

### PR DESCRIPTION
**Pull request by UdhayaShan1**
`EditCommandParser` will now return stacked error messages to make it clearer to users which input went wrong.

Refactor `EditCommandParser` and `AddCommandParser` to adhere to better code quality standards.

Add comprehensive testing for `EditCommand`.

### CODE REVIEW

File: src/main/java/seedu/realodex/logic/parser/AddCommandParser.java [hunk 1]
```diff
@@ -37,106 +41,157 @@ public class AddCommandParser implements Parser<AddCommand> {
...
-        phoneStored.buildErrorMessage(errorMessageBuilder, "phone");
+        if (errorMessageBuilder.length() > 0) {
+            throw new ParseException(errorMessageBuilder.toString().trim());
+        }
+        Birthday birthday = birthdayStored.returnStoredResult();
 
-        ParserUtilResult<Income> incomeStored =
-                ParserUtil.parseIncomeReturnStored(argMultimap.getValue(PREFIX_INCOME).orElseThrow());
+        return new Person(name, phone, income, email, address, family, tags, housingType, remark, birthday);
+    }
```
Comment (appleraincoat) [inline src/main/java/seedu/realodex/logic/parser/AddCommandParser.java:132]: 👍

File: src/test/java/seedu/realodex/logic/parser/EditCommandParserTest.java [hunk 2]
```diff
@@ -112,34 +112,80 @@ public void parse_invalidPreamble_failure() {
...
+    public void parse_invalidMultipleValues_failure() {
+        String errorMessage = "Error parsing %s: ";
+
+        // Test case 1: Invalid name, phone, and email
+        String userInput1 = "1" + INVALID_NAME_DESC + INVALID_PHONE_DESC
+                + INVALID_EMAIL_DESC;
+
+        String expectedErrorMessage1 = String.format(errorMessage, "name") + Name.MESSAGE_CONSTRAINTS + "\n"
+                + String.format(errorMessage, "phone") + Phone.MESSAGE_CONSTRAINTS + "\n"
+                + String.format(errorMessage, "email") + Email.MESSAGE_CONSTRAINTS + "\n";
```
Comment (appleraincoat) [inline src/test/java/seedu/realodex/logic/parser/EditCommandParserTest.java:163]: sheeeesh 👍

File: src/test/java/seedu/realodex/logic/commands/EditCommandTest.java [hunk 3]
```diff
@@ -148,6 +160,212 @@ public void execute_invalidPersonIndexFilteredList_failure() {
...
+        EditCommand editCommand = new EditCommand(indexLastPerson, descriptor);
+
+        Model expectedModel = new ModelManager(new Realodex(model.getRealodex()), new UserPrefs());
+        expectedModel.setPerson(lastPerson, editedPerson);
+        String expectedMessage = String.format(EditCommand.MESSAGE_EDIT_PERSON_SUCCESS, Messages.format(editedPerson));
+
+        assertCommandSuccess(editCommand, model, expectedMessage, expectedModel);
+    }
+
+
```
Comment (appleraincoat) [inline src/test/java/seedu/realodex/logic/commands/EditCommandTest.java:368]: great job on the tests here, very exhaustive and comprehensive

**Review by appleraincoat — APPROVED** [review]
LGTM! Nice work in such short notice.

File: src/main/java/seedu/realodex/logic/parser/AddCommandParser.java [hunk 4]
```diff
@@ -37,106 +41,157 @@ public class AddCommandParser implements Parser<AddCommand> {
...
+     * @param argMultimap An ArgumentMultimap containing the parsed arguments.
+     * @throws ParseException If any compulsory prefix is missing.
+     */
+    private void validateCompulsoryPrefixes(ArgumentMultimap argMultimap) throws ParseException {
+        Prefix[] compulsoryPrefixes = {PREFIX_NAME, PREFIX_ADDRESS, PREFIX_INCOME, PREFIX_PHONE,
+            PREFIX_FAMILY, PREFIX_EMAIL,
+            PREFIX_TAG, PREFIX_HOUSINGTYPE};
+        if (!arePrefixesPresent(argMultimap, compulsoryPrefixes)) {
+            String missingPrefixesMessage = argMultimap.returnMessageOfMissingPrefixes(compulsoryPrefixes) + "\n";
+            throw new ParseException(Messages.getErrorMessageForMissingPrefixes(missingPrefixesMessage)
```
Comment (4llysa) [inline src/main/java/seedu/realodex/logic/parser/AddCommandParser.java:85]: Could you make the parameters all on the same line?
[BIRTHDAY] should be on the same line as the rest
image

Comment (4llysa) [reply src/main/java/seedu/realodex/logic/parser/AddCommandParser.java:85]: Perhaps Parameters should start in a new line too

**Review by 4llysa — CHANGES_REQUESTED** [review]

File: src/main/java/seedu/realodex/logic/parser/EditCommandParser.java [hunk 5]
```diff
@@ -34,54 +42,79 @@ public EditCommand parse(String args) throws ParseException {
...
+
         if (!editPersonDescriptor.isAnyFieldEdited()) {
             throw new ParseException(EditCommand.MESSAGE_NOT_EDITED);
         }
 
-        return new EditCommand(index, editPersonDescriptor);
+        return editPersonDescriptor;
+    }
+
+    private <T> void parseAndSetField(ArgumentMultimap argMultimap, Prefix prefix, Consumer<T> setter,
```
Comment (4llysa) [inline src/main/java/seedu/realodex/logic/parser/EditCommandParser.java:110]: good abstraction

File: src/test/java/seedu/realodex/model/person/BirthdayTest.java [hunk 6]
```diff
@@ -56,6 +56,7 @@ public void isValidBirthday_test() {
         assertTrue(Birthday.isValidBirthday("29Feb2024")); // leap year
         assertTrue(Birthday.isValidBirthday("12May2003"));
         assertTrue(Birthday.isValidBirthday("08Aug1888"));
+        assertTrue(Birthday.isValidBirthday("08Aug888")); //before year 1000
```
Comment (4llysa) [inline src/test/java/seedu/realodex/model/person/BirthdayTest.java:59]: remove this birthday does not support before year 1000

File: src/test/java/seedu/realodex/testutil/TypicalPersons.java [hunk 7]
```diff
@@ -87,6 +87,42 @@ public class TypicalPersons {
...
+
+    public static final Person NATHANIEL = new PersonBuilder().withName("Nathaniel Martinez").withPhone("9482891")
+            .withIncome("140000").withEmail("nathaniel@example.com").withAddress("elm street").withTags("seller")
+            .withHousingType("HDB").withBirthday("14Aug1972").build();
+
+    public static final Person OLIVIA = new PersonBuilder().withName("Olivia Taylor").withPhone("9482889")
+            .withIncome("150000").withEmail("olivia@example.com").withAddress("maple street").withTags("buyer")
+            .withHousingType("HDB").withBirthday("17Dec1968").build();
+
+
```
Comment (4llysa) [inline src/test/java/seedu/realodex/testutil/TypicalPersons.java:122]: did u add NEW PEOPLE

**Review by 4llysa — CHANGES_REQUESTED** [review]
thanks for adding. just remove the birthday test for <1000 year

**Comment by 4llysa** [comment]
can check if the default persons from realodex.json get loaded up properly?

File: src/main/java/seedu/realodex/logic/parser/AddCommandParser.java (hunk 4 above)
Comment (UdhayaShan1) [reply src/main/java/seedu/realodex/logic/parser/AddCommandParser.java:85]: image

Thanks

File: src/test/java/seedu/realodex/testutil/TypicalPersons.java (hunk 7 above)
Comment (UdhayaShan1) [reply src/test/java/seedu/realodex/testutil/TypicalPersons.java:122]: My friend gpt did 😊

Comment (UdhayaShan1) [reply src/test/java/seedu/realodex/testutil/TypicalPersons.java:122]: I wanted to test sorting using another new list

Comment (4llysa) [reply src/test/java/seedu/realodex/testutil/TypicalPersons.java:122]: 👍🏻

**Review by 4llysa — APPROVED** [review]
LGTM!

**Pull request merged**